applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_base_element.cpp

```diff
@@ -555,19 +555,33 @@ void AdjointFiniteDifferencingBaseElement<TPrimalElement>::CalculateStressDesign
 
 // private
 template <class TPrimalElement>
-double AdjointFiniteDifferencingBaseElement<TPrimalElement>::GetPerturbationSize(const Variable<double>& rDesignVariable)
+double AdjointFiniteDifferencingBaseElement<TPrimalElement>::GetPerturbationSize(const Variable<double>& rDesignVariable, const ProcessInfo& rCurrentProcessInfo)
 {
-    const double correction_factor = this->GetPerturbationSizeModificationFactor(rDesignVariable);
-    const double delta = this->GetValue(PERTURBATION_SIZE) * correction_factor;
+    double delta = 0.0;
+    if(rCurrentProcessInfo[ADAPT_PERTURBATION_SIZE])
+    {
+        const double correction_factor = this->GetPerturbationSizeModificationFactor(rDesignVariable);
+        delta = rCurrentProcessInfo[PERTURBATION_SIZE] * correction_factor;
+    }
+    else
+        delta = rCurrentProcessInfo[PERTURBATION_SIZE];
+
     KRATOS_DEBUG_ERROR_IF_NOT(delta > 0) << "The perturbation size is not > 0!";
     return delta;
 }
 
 template <class TPrimalElement>
-double AdjointFiniteDifferencingBaseElement<TPrimalElement>::GetPerturbationSize(const Variable<array_1d<double,3>>& rDesignVariable)
+double AdjointFiniteDifferencingBaseElement<TPrimalElement>::GetPerturbationSize(const Variable<array_1d<double,3>>& rDesignVariable, const ProcessInfo& rCurrentProcessInfo)
 {
-    const double correction_factor = this->GetPerturbationSizeModificationFactor(rDesignVariable);
-    const double delta = this->GetValue(PERTURBATION_SIZE) * correction_factor;
+    double delta = 0.0;
+    if(rCurrentProcessInfo[ADAPT_PERTURBATION_SIZE])
+    {
+        const double correction_factor = this->GetPerturbationSizeModificationFactor(rDesignVariable);
+        delta = rCurrentProcessInfo[PERTURBATION_SIZE] * correction_factor;
+    }
+    else
+        delta = rCurrentProcessInfo[PERTURBATION_SIZE];
+
```

Comment:
recommendation:
~~~c
double delta = rCurrentProcessInfo[PERTURBATION_SIZE];
if (rCurrentProcessInfo[ADAPT_PERTURBATION_SIZE]) {
        delta *= this->GetPerturbationSizeModificationFactor(rDesignVariable);
}
~~~